acq4/devices/PatchPipette/states.py

```diff
@@ -1130,97 +1163,219 @@ class PatchPipetteResealState(PatchPipetteState):
+        'retractionSuccessDistance': {'type': 'float', 'default': 200e-6, 'suffix': 'm'},
+        'retractionSuccessState': {'type': 'str', 'default': 'home with nucleus'},
+        'detectionTau': {'type': 'float', 'default': 1, 'suffix': 's'},
+        'repairTau': {'type': 'float', 'default': 10, 'suffix': 's'},
+        'stretchDetectionRatio': {'type': 'float', 'default': 1.05},
+        'stretchRecoveryTime': {'type': 'float', 'default': 30, 'suffix': 's'},
+        'tearDetectionRatio': {'type': 'float', 'default': 1},
+        'tearRecoveryTime': {'type': 'float', 'default': 60, 'suffix': 's'},
     }
 
     def __init__(self, *args, **kwds):
-        self.retractionFuture = None
         PatchPipetteState.__init__(self, *args, **kwds)
+        self._retractionFuture = None
+        self._pressureFuture = None
+        self._startPosition = np.array(self.dev.pipetteDevice.globalPosition())
+        self._detectionResistanceRatio = 1
+        self._repairResistanceRatio = 1
+        self._lastTestPulseTime = None
+        self._lastTestPulseSSResist = None
+
+    def nuzzle(self):
+        """Wiggle the pipette around inside the cell to clear space for a nucleus to be extracted."""
+        self.setState("nuzzling")
+        if self._retractionFuture is not None:
+            self._retractionFuture.stop()
+        if self._pressureFuture is not None:
+            self._pressureFuture.stop()
+        pos = np.array(self.dev.pipetteDevice.globalPosition())
+        # TODO move back a little?
+        pipette_direction = self.dev.pipetteDevice.globalDirection()
+
+        def random_wiggle_direction():
+            """pick a random point on a circle perpendicular to the pipette axis"""
+            while np.linalg.norm(vec := np.cross(pipette_direction, np.random.uniform(-1, 1, size=3))) == 0:
+                pass  # prevent division by zero
+            return self.config['nuzzleLateralWiggleRadius'] * vec / np.linalg.norm(vec)
+
+        prev_dir = random_wiggle_direction()
+        for _ in range(self.config['nuzzleRepetitions']):
+            self.dev.pressureDevice.setPressure(source='regulator', pressure=self.config['nuzzleInitialPressure'])
+            self._pressureFuture = self.dev.pressureDevice.rampPressure(
+                target=self.config['nuzzlePressureLimit'], duration=self.config['nuzzleDuration'])
+            start = ptime.time()
+            while ptime.time() - start < self.config['nuzzleDuration']:
+                while np.dot(direction := random_wiggle_direction(), prev_dir) > 0:
+                    pass  # ensure different direction from previous
+                self.waitFor(self.dev.pipetteDevice._moveToGlobal(pos=pos + direction, speed=self.config['nuzzleSpeed']))
+                prev_dir = direction
+            self.waitFor(self.dev.pipetteDevice._moveToGlobal(pos=pos, speed=self.config['nuzzleSpeed']))
+            self.waitFor(self._pressureFuture)
+
+    def handleTear(self):
```

Comment:
Maybe we don't need handleTear / handleStretch / tearRecoveryTime / stretchRecoveryTime ?